plutus-tx/src/PlutusTx/Blueprint/Contract.hs

```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+module PlutusTx.Blueprint.Contract where
+
+import Prelude
+
+import Data.Aeson (ToJSON (..), (.=))
+import Data.Aeson qualified as Aeson
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Maybe (catMaybes)
+import Data.Text (Text)
+import PlutusTx.Blueprint.Definition (DefinitionId)
+import PlutusTx.Blueprint.Preamble (Preamble)
+import PlutusTx.Blueprint.Schema (Schema)
+import PlutusTx.Blueprint.Validator (ValidatorBlueprint)
+
+data ContractBlueprint = MkContractBlueprint
+  { contractId          :: Maybe Text
+  -- ^ An optional identifier for the contract
+  , contractPreamble    :: Preamble
+  -- ^ An object with meta-information about the contract
+  , contractValidators  :: [ValidatorBlueprint]
+  -- ^ An object of named validators
+  , contractDefinitions :: Map DefinitionId Schema
+  -- ^ A registry of definition re-used across the specification
+  }
+  deriving stock (Show)
+
+instance ToJSON ContractBlueprint where
+  toJSON MkContractBlueprint{..} =
+    Aeson.object
+      $ catMaybes
+        [ Just $ "$schema" .= schemaUrl
+        , fmap ("$id" .=) contractId
+        , Just
+            $ "$vocabulary"
+            .= Aeson.object
+              [ "https://json-schema.org/draft/2020-12/vocab/core" .= True
+              , "https://json-schema.org/draft/2020-12/vocab/applicator" .= True
+              , "https://json-schema.org/draft/2020-12/vocab/validation" .= True
+              , "https://cips.cardano.org/cips/cip57" .= True
+              ]
+        , Just $ "preamble" .= contractPreamble
+        , Just $ "validators" .= contractValidators
+        , if Map.null contractDefinitions
```

Comment:
again, if we wanted to be generic we could write a newtype wrapper around Map that omits it when it is empty.

Reply:
In which case Blueprint API users will have to know about such a map wrapper and will have to deal with it. I think this price for the automatic derivation is not justified.